R/plot-build.R

```diff
@@ -481,3 +398,132 @@ table_add_tag <- function(table, label, theme) {
+
+    position <- theme$legend.position.inside %||% just
+    legends$inside <- editGrob(
+      legends$inside,
+      vp = viewport(
+        x = position[1],
+        y = position[2],
+        just = just,
+        height = heights$inside,
+        width  = widths$inside
+      )
+    )
+  }
+
+  legends[.trbl] <- lapply(.trbl, function(position) {
+    box <- .subset2(legends, position)
+    if (is.zero(box)) {
+      return(box)
+    }
+    just <- paste0("legend.justification.", position)
+    just <- valid.just(calc_element(just, theme))
+
+    # First value is xjust, second value is yjust
+    xjust <- just[1]
+    yjust <- just[2]
+
+    margin <- margin(1 - yjust, 1 - xjust, yjust, xjust, unit = "null")
+
+    vp <- viewport(
+      x = xjust, y = yjust, just = just,
+      height = .subset2(heights, position),
+      width  = .subset2(widths,  position)
+    )
+    box <- editGrob(box, vp = vp)
+    gtable_add_padding(box, margin)
+  })
+
+  # If legend is missing, set spacing to zero for that legend
+  zero    <- unit(0, "pt")
+  spacing <- lapply(empty, function(is_empty) if (is_empty) zero else spacing)
+
+  location <- switch(
+    theme$legend.location %||% "panel",
+    "plot" = plot_extent,
+    find_panel
+  )
+
+  place <- location(table)
+
+  # Add right legend
+  table <- gtable_add_cols(table, spacing$right, pos = -1)
+  table <- gtable_add_cols(table, widths$right,  pos = -1)
+  table <- gtable_add_grob(
+    table, legends$right, clip = "off",
+    t = place$t, b = place$b, l = -1, r = -1,
+    name = "guide-box-right"
+  )
+
+  # Add left legend
+  table <- gtable_add_cols(table, spacing$left, pos = 0)
+  table <- gtable_add_cols(table, widths$left,  pos = 0)
+  table <- gtable_add_grob(
+    table, legends$left, clip = "off",
+    t = place$t, b = place$b, l = 1, r = 1,
+    name = "guide-box-left"
+  )
+
+  place <- location(table)
+
+  # Add bottom legend
+  table <- gtable_add_rows(table, spacing$bottom, pos = -1)
+  table <- gtable_add_rows(table, heights$bottom, pos = -1)
+  table <- gtable_add_grob(
+    table, legends$bottom, clip = "off",
+    t = -1, b = -1, l = place$l, r = place$r,
+    name = "guide-box-bottom"
+  )
+
+  # Add top legend
+  table <- gtable_add_rows(table, spacing$top, pos = 0)
+  table <- gtable_add_rows(table, heights$top, pos = 0)
+  table <- gtable_add_grob(
+    table, legends$top, clip = "off",
+    t = 1, b = 1, l = place$l, r = place$r,
+    name = "guide-box-top"
+  )
+
+  # Add manual legend
+  place <- find_panel(table)
+  table <- gtable_add_grob(
+    table, legends$inside, clip = "off",
+    t = place$t, b = place$b, l = place$l, r = place$r,
+    name = "guide-box-inside"
+  )
+
+  table
+}
+
+plot_extent <- function(table) {
+  layout <- table$layout
+  data_frame0(
+    t = min(.subset2(layout, "t")),
+    r = max(.subset2(layout, "r")),
+    b = max(.subset2(layout, "b")),
+    l = min(.subset2(layout, "l")),
```

Comment:
Same with this use of `.subset2()`. If there is a good reason, fine. Otherwise use standard indexing